subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/transform/TransformationChain.java

```diff
@@ -18,27 +18,51 @@
 
 import org.gradle.api.Action;
 
+import javax.annotation.Nullable;
+
 /**
  * A series of {@link TransformationStep}s.
  */
 public class TransformationChain implements Transformation {
 
-    private final Transformation first;
-    private final Transformation second;
+    private final Transformation init;
+    private final TransformationStep last;
     private final int stepsCount;
 
-    public TransformationChain(Transformation first, Transformation second) {
-        this.first = first;
-        this.second = second;
-        this.stepsCount = first.stepsCount() + second.stepsCount();
+    private static final Transformation EMPTY = new Transformation() {
+        @Override
+        public int stepsCount() {
+            return 0;
+        }
+
+        @Override
+        public boolean requiresDependencies() {
+            return false;
+        }
+
+        @Override
+        public void visitTransformationSteps(Action<? super TransformationStep> action) {
+        }
+
+        @Override
+        public String getDisplayName() {
+            return "EMPTY";
+        }
+    };
+
+    public TransformationChain(@Nullable Transformation init, TransformationStep last) {
+        this.init = init == null ? EMPTY : init;
+        this.last = last;
+        this.stepsCount = this.init.stepsCount() + 1;
     }
 
-    public Transformation getFirst() {
-        return first;
+    @Nullable
+    public Transformation getInit() {
```

Comment:
❔ Should we add a javadoc that explains what `init` is? 

It may not be trivial without reading the code that this is the chain without the last step.

Reply:
I added some Javadoc. FWIW, I took the names from "regular" functional programming. You have `head`/`tail` and similar `init`/`last`.